src/node.rs

```diff
@@ -346,6 +236,85 @@ impl<K: ArtKey, V: Default, const MAX_PARTIAL_LEN: usize> ArtNode<K, V, MAX_PART
         }
     }
 
+    #[inline]
+    fn compression(&mut self, prefix_mismatch_pos: usize, key: K, depth: usize, val: V) {
+        let mut old_node = std::mem::replace(self, ArtNode::node4(Box::default()));
+        // self is already new node
+
+        let old_node_header = old_node.header();
+        let new_node_header = self.header_mut();
+
+        // copy matched partial from old node to new node.
+        let max_copy_len = min(prefix_mismatch_pos, MAX_PARTIAL_LEN);
+        new_node_header.partial.data[..max_copy_len]
+            .copy_from_slice(&old_node_header.partial.data[..max_copy_len]);
+        new_node_header.partial.len = prefix_mismatch_pos as u32;
+
+        // Note: The match may exceed the maximum of the vector store, so take the minimum of both.
+        // Artful uses mixed compression, So the actual storage length of partial can exceed the
+        // maximum value of the vector.
+        // pessimistic compression
+        if old_node_header.partial.len as usize <= MAX_PARTIAL_LEN {
+            let old_node_header = old_node.header_mut();
+            let old_node_byte = old_node_header.partial.data[prefix_mismatch_pos];
+            old_node_header.partial.len -= (prefix_mismatch_pos + 1) as u32;
+            old_node_header
+                .partial
+                .data
+                .rotate_left(prefix_mismatch_pos + 1);
+            // unsafe {
+            //     std::ptr::copy(
+            //         old_node_header
+            //             .partial
+            //             .data
+            //             .as_ptr()
+            //             .offset(prefix_mismatch_pos as isize + 1),
+            //         old_node_header.partial.data.as_mut_ptr(),
+            //         min(MAX_PARTIAL_LEN, old_node_header.partial.len as usize),
+            //     );
+            // }
```

Comment:
use `rotate_left`  instead of unsafe code is workable，such as common 
prefix is `a | b | c |` abd  `old_node` prefix is `a | b | c | d | e `, `rotate_left` after get `e| a | b | c | d ` and length of partial update to 1.

but if `rotate_left`  length equal to partial length, to call `rotate_left`  is no-op.  this boundary condition produces a result that is correct because the partial length is set correctly.